### PR TITLE
Gamma stage

### PR DESCRIPTION
**Pull request by reillytilbury**
- Refactored UX into spec.py and config.yaml files
- Removed some verbose checking functions
- Removed diagnostics module for plotter
- removed options for param estimator with image feedback
- Set new command for setting up stubs in spec and config
- Rerouted a lot of error handling outside the hypothesis engine code
- Better documentation

### CODE REVIEW

File: src/cli.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+
+
+REQUIRED_SPEC_FUNCTIONS = [
+    "load_and_process_data",
+    "train_test_split",
+    "model_v1",
+    "param_est_v1",
+    "model_v2",
+    "param_est_v2",
+]
```
[the remark this answers is not in the document]
Comment (dabin-k) [reply src/cli.py:16]: Let's add loss_fn in REQUIRED_SPEC_FUNCTIONS and remove the option to specify the loss_fn in the yaml config

File: run.py [hunk 2]
```diff
@@ -25,114 +37,365 @@ def deep_merge(base: dict, override: dict) -> dict:
...
-        data_module = importlib.import_module(module_path)
-        create_train_test_trial_split_fn = getattr(data_module, function_name)
-    else:
-        create_train_test_trial_split_fn = None
+    # Image diagnostics are enabled automatically if spec defines plot_model_fits.
+    spec_plot_fn = getattr(spec_module, "plot_model_fits", None)
+    plot_model_fits_fn = _build_plot_model_fits_fn(spec_plot_fn) if callable(spec_plot_fn) else None
 
-    loss_fn_path = config.pop('loss_fn', None)
+    # Optional config-level loss override still supported.
```
Comment (dabin-k) [inline run.py]: Let's delete this optional config-level loss override

File: src/prompt_test.py [hunk 3]
```diff
@@ -302,7 +302,6 @@ def test_experiment_params_types(self):
             'exploit_point': float,
             'param_penalty_weight': float,
             'FAILED_PROGRAM_COST': float,  # Must be float, not string!
-            'use_image_feedback': bool,
             'use_param_estimator': bool,
```
[the remark this answers is not in the document]
Comment (dabin-k) [reply src/prompt_test.py:305]: All projects updated

File: src/hypothesis_engine.py [hunk 4]
```diff
@@ -1159,28 +270,23 @@ def objective_vectorized(model, param_estimator, x, y, create_train_test_trial_s
...
-    # Split trials for inputs and outputs (may use different indices when mismatched)
-    x_train_trial_idx, x_test_trial_idx, y_train_trial_idx, y_test_trial_idx = \
-        _call_trial_split(create_train_test_trial_split_fn, n_trials_x, n_trials_y, random_seed)
-
-    # Split inputs and outputs using their respective trial indices
-    x_train = x_data[:, :, x_train_trial_idx]  # (n_samples, n_features, n_train_trials_x)
-    y_train = y_data[:, :, y_train_trial_idx]   # (n_samples, n_targets, n_train_trials_y)
-    x_test = x_data[:, :, x_test_trial_idx]     # (n_samples, n_features, n_test_trials_x)
-    y_test = y_data[:, :, y_test_trial_idx]      # (n_samples, n_targets, n_test_trials_y)
+    if loss_fn is None:
```
Comment (dabin-k) [inline src/hypothesis_engine.py:274]: Do we want this to be the default behaviour? Maybe we should just raise an error if not specified?

File: projects/grid_cells/config.yaml
Comment (dabin-k) [inline projects/grid_cells/config.yaml:1]: I'm going to assume that all the updated configs have been generated in a principled way. I didn't read through the project files

File: edgar.py [hunk 5]
```diff
@@ -0,0 +1,5 @@
+from src.cli import main
```
Comment (dabin-k) [inline edgar.py]: We should rename this file?

File: requirements.txt [hunk 6]
```diff
@@ -1,7 +1,7 @@
 anthropic==0.50.0
 google-genai==1.51.0
-jax==0.7.1
-jaxlib==0.7.1
+jax==0.4.8
+jaxlib==0.4.38
```
[the remark this answers is not in the document]
Comment (dabin-k) [reply requirements.txt:4]: @reillytilbury Do you know why you have different versions here?

File: run.py [hunk 7]
```diff
@@ -25,114 +37,365 @@ def deep_merge(base: dict, override: dict) -> dict:
...
+def _broadcast_model_loss(loss_value, n_samples: int):
+    if loss_value is None:
+        return None
+    loss_arr = np.asarray(loss_value)
+    if loss_arr.size == 0:
+        return None
+    if loss_arr.ndim == 0:
+        return np.full(n_samples, float(loss_arr))
+    flat = loss_arr.reshape(-1)
+    if flat.size == 1:
```
Comment (dabin-k) [inline run.py:176]: When does this if case hold?

File: run.py [hunk 8]
```diff
@@ -25,114 +37,365 @@ def deep_merge(base: dict, override: dict) -> dict:
...
+        return np.full(n_samples, float(flat[0]))
+    if flat.size != n_samples:
+        return None
```
Comment (dabin-k) [inline run.py]: Do we want to return None here or an error?

File: run.py [hunk 9]
```diff
@@ -25,114 +37,365 @@ def deep_merge(base: dict, override: dict) -> dict:
...
+            call_kwargs = {k: v for k, v in call_kwargs.items() if k in allowed_kwargs}
+        spec_plot_fn(**call_kwargs)
+
+    return _wrapped_plot_fn
 
 
 async def _run_many(test_mode: bool = False, config_path: str = "config.yaml"):
+    """
+    _run_many(test_mode: bool = False, config_path: str = "config.yaml")
+    Asynchronously runs multiple experiments based on the provided configuration and parameters.
```
Comment (dabin-k) [inline run.py:279]: false?

File: run.py [hunk 10]
```diff
@@ -146,11 +418,49 @@ async def _run_many(test_mode: bool = False, config_path: str = "config.yaml"):
...
+
         print("running with standard params")
-        await hypothesis_engine.hypothesis_engine(
+        full_dir = await hypothesis_engine.hypothesis_engine(
             n_iterations=params['n_iterations'],
             time_limit=params['time_limit'],
-            use_image_feedback=params['use_image_feedback'],
             use_large_every=params['use_large_every'],
             param_penalty_weight=params['param_penalty_weight'],
             exploration_topology=params['exploration_topology'],
```
[the remark this answers is not in the document]
Comment (dabin-k) [reply run.py:455]: We should add the exploitation_topology!

File: src/hypothesis_engine.py [hunk 11]
```diff
@@ -1388,488 +503,148 @@ def eval_loss_batched(params_2d, x_eval, y_eval):
...
-    Select evaluation points for model diagnostics.
+    programs_list = []
+    if programs_df is None or len(programs_df) == 0:
+        return programs_list
+    x_arr = jnp.asarray(ensure_inputs(x).to_tensor())
+    y_arr = jnp.asarray(ensure_outputs(y).to_tensor())
+    n_samples = x_arr.shape[0]
+
+    if loss_func is None:
+        loss_func = lambda y_pred, y_true: (y_true - y_pred) ** 2
```
Comment (dabin-k) [inline src/hypothesis_engine.py]: We should just error here?

File: src/prompt_manager.py [hunk 12]
```diff
@@ -66,32 +108,33 @@ def get_program_prompt_legacy(self, programs_df : pd.DataFrame, mode : str, use_
...
+                model_idx=f"{model_idx}",
+                train_loss=self._format_loss(train_loss),
+                program_code_string=program_code_string,
+            )
+            prompt += "\n\n" + per_model_prompt
         
         return prompt
     
-    def get_parameter_estimator_prompt_legacy(self, programs_df : pd.DataFrame, model_code_string : str, max_lines : int = 100, use_image : bool = True) -> str:
+    def get_parameter_estimator_prompt_legacy(self, programs_df : pd.DataFrame, model_code_string : str, max_lines : int = 100) -> str:
```
Comment (dabin-k) [inline src/prompt_manager.py]: Why is this called get_parameter_estimator_prompt_legacy? Should we rename it if it's not legacy?

File: tests/test_objective_function.py [hunk 13]
```diff
@@ -81,6 +81,20 @@ def simple_param_estimator(X, y):
...
+def default_loss_fn(model, x_i, y_i, params):
+    """Default per-sample loss: MSE averaged over all outputs and trials.
+
+    Works for both scalar outputs (n_trials,) and vectorized outputs (n_targets, n_trials).
+    Equivalent to uniform-weight MSE across all targets and trials.
+    """
+    pred = model(x_i, *params)
+    if pred.ndim == 1:
+        pred = pred[None, :]  # normalize to (1, n_trials)
+    return jnp.mean((pred - y_i) ** 2)
```
[the remark this answers is not in the document]
Comment (dabin-k) [reply tests/test_objective_function.py]: This test file was helpful when I was updating the objective function a lot, but it's now more effort to update it. It should be fine to just delete this file - I don't think any of the tests are that valuable anymore

File: tests/test_orientation_tuning_seed_loss_regression.py [hunk 14]
```diff
@@ -69,7 +68,6 @@ def _compute_seed_losses(settings):
         _, _, train_loss, _ = objective(
             model=model,
             param_estimator=estimator,
-            loss_func=quadratic_loss,
             x=inputs_train,
```
[the remark this answers is not in the document]
Comment (dabin-k) [reply tests/test_orientation_tuning_seed_loss_regression.py]: Probably also fine to remove this test file - it doesn't add much atm and the current version is broken

**Review by dabin-k — COMMENTED** [review]
Great stuff! I left some comments which would benefit from being addressed but nothing major.

File: requirements.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (reillytilbury) [reply requirements.txt:4]: This was the only configuration I could get to work on my Mac

File: run.py (hunk 8 above)
Comment (reillytilbury) [reply run.py]: Error! Changing now!

File: src/prompt_manager.py (hunk 12 above)
Comment (reillytilbury) [reply src/prompt_manager.py]: I think this was called this when chat mode was introduced and old things became known as legacy. But since chat mode is not default I think we should just call those other funcs ..._chat. I am resolving now

**Comment by dabin-k** [comment]
Let's delete this branch? Going forward, let's give our branches meaningful names as well!

File: requirements.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (dabin-k) [reply requirements.txt:4]: In the current state, other people won't be able to setup a conda environment using this requirements.txt. I'm going to change the requirement file so that it's 0.4.38 for both jax and jaxlib in that case.
